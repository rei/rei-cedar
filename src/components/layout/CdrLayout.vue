<script setup lang="ts">
import { useCssModule, computed } from 'vue';
import mapClasses from '../../utils/mapClasses';
import { getStructureStyles } from '../../utils/layout';
import type { Layout, NameValuePair } from '../../types/interfaces';
import type { Structure } from '../../types/other';
import { modifyClassName } from '../../utils/buildClass';

/** Foundational container for creating structured layouts */

defineOptions({ name: 'CdrLayout' });

const props = withDefaults(defineProps<Layout>(), {
  gap: 'zero',
  rowGap: 'zero',
  columnGap: 'zero',
  columns: undefined,
  rows: undefined,
  as: 'div',
  queryType: 'container',
  flow: undefined,
  flowValue: 'auto',
});

const style = useCssModule();

const rootProps = computed(() => {
  const baseClass = 'cdr-layout';
  const classes = [baseClass];
  const inlineStyles: NameValuePair = {};

  // Add gap for entire grid
  if (props.gap !== 'zero') {
    classes.push(modifyClassName(baseClass, `gap-${props.gap}`));
  }

  // Add gap for columns
  if (props.columnGap !== 'zero') {
    classes.push(modifyClassName(baseClass, `column-gap-${props.columnGap}`));
  }

  // Add gap for rows
  if (props.rowGap !== 'zero') {
    classes.push(modifyClassName(baseClass, `row-gap-${props.rowGap}`));
  }

  // Add grid-auto-flow and supporting value
  if (props.flow) {
    classes.push(modifyClassName(baseClass, `flow-${props.flow}`));
    inlineStyles['--cdr-layout-flow-value'] = props.flowValue;
  }

  // Add grid templates for columns and rows
  ['rows' as Structure, 'columns' as Structure].forEach((structure) => {
    if (!props[structure]) {
      return;
    }

    // Add class for applying inline grid styles for rows or columns
    classes.push(modifyClassName(baseClass, structure));

    // Add responsive styles if it is the type StructureObject
    if (typeof props[structure] === 'object' && !Array.isArray(props[structure])) {
      const querySuffix = props.queryType === 'container' ? 'cq' : 'mq';
      classes.push(modifyClassName(baseClass, `${structure}-${querySuffix}`));
    }

    // Add inline styles that will be picked up by classes
    const structureStyles = getStructureStyles({ props, structure });
    Object.assign(inlineStyles, structureStyles);
  });

  return { class: mapClasses(style, ...classes) || undefined, style: inlineStyles };
});
</script>

<template>
  <component
    :is="props.as"
    v-bind="rootProps"
  >
    <!-- @slot Where all default content should be placed. -->
    <slot />
  </component>
</template>

<style lang="scss" module src="./styles/CdrLayout.module.scss"></style>
