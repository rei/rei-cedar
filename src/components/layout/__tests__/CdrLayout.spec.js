import { mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrLayout from '../CdrLayout.vue';

describe('CdrLayout', () => {
  it('default snapshot', () => {
    const wrapper = mount(CdrLayout);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('sets the default tag prop correctly', () => {
    const wrapper = mount(CdrLayout, { props: { as: 'span' } });
    expect(wrapper.element.tagName).toBe('SPAN');
  });

  it('base classes only', () => {
    const wrapper = mount(CdrLayout, { props: { as: 'span' } });
    expect(wrapper.classes()).toEqual(['cdr-layout']);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('shows default slot', () => {
    const wrapper = mount(CdrLayout, { slots: { default: 'Default slot' } });
    expect(wrapper.find('.cdr-layout').text()).toBe('Default slot');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('basic columns', () => {
    const wrapper = mount(CdrLayout, { props: { columns: 2 } });
    expect(wrapper.classes()).toEqual(['cdr-layout', 'cdr-layout--columns']);
    expect(wrapper.find('.cdr-layout').attributes('style')).toBe('--cdr-layout-columns: 1fr 1fr;');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('ratio columns', () => {
    const wrapper = mount(CdrLayout, { props: { columns: [2, 1] } });
    expect(wrapper.classes()).toEqual(['cdr-layout', 'cdr-layout--columns']);
    expect(wrapper.find('.cdr-layout').attributes('style')).toBe('--cdr-layout-columns: 2fr 1fr;');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('flow to generate columns', () => {
    const wrapper = mount(CdrLayout, {
      props: { flow: 'column', flowValue: '100px' },
      slots: { default: '<div>1</div><div>2</div><div>3</div>' },
    });
    expect(wrapper.classes()).toEqual(['cdr-layout', 'cdr-layout--flow-column']);
    expect(wrapper.find('.cdr-layout').attributes('style')).toBe('--cdr-layout-flow-value: 100px;');
    expect(wrapper.findAll('.cdr-layout div').length).toBe(3);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('container query columns', () => {
    const wrapper = mount(CdrLayout, {
      props: { columns: { xs: 1, sm: ['100px', 1], md: ['400px', 1], lg: ['800px', 1] } },
    });
    expect(wrapper.classes()).toEqual([
      'cdr-layout',
      'cdr-layout--columns',
      'cdr-layout--columns-cq',
    ]);
    expect(wrapper.find('.cdr-layout').attributes('style')).toBe(
      '--cdr-layout-columns-xs: 1fr; --cdr-layout-columns-sm: 100px 1fr; --cdr-layout-columns-md: 400px 1fr; --cdr-layout-columns-lg: 800px 1fr;',
    );
    expect(wrapper.element).toMatchSnapshot();
  });

  it('media query columns', () => {
    const wrapper = mount(CdrLayout, {
      props: {
        queryType: 'media',
        columns: { xs: 1, sm: ['100px', 1], md: ['400px', 1], lg: ['800px', 1] },
      },
    });
    expect(wrapper.classes()).toEqual([
      'cdr-layout',
      'cdr-layout--columns',
      'cdr-layout--columns-mq',
    ]);
    expect(wrapper.find('.cdr-layout').attributes('style')).toBe(
      '--cdr-layout-columns-xs: 1fr; --cdr-layout-columns-sm: 100px 1fr; --cdr-layout-columns-md: 400px 1fr; --cdr-layout-columns-lg: 800px 1fr;',
    );
    expect(wrapper.element).toMatchSnapshot();
  });

  it('same gap', () => {
    const wrapper = mount(CdrLayout, { props: { columns: 2, gap: 'one-x' } });
    expect(wrapper.classes()).toEqual([
      'cdr-layout',
      'cdr-layout--gap-one-x',
      'cdr-layout--columns',
    ]);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('different gap', () => {
    const wrapper = mount(CdrLayout, {
      props: { columns: 2, columnGap: 'one-x', rowGap: 'one-x' },
    });
    expect(wrapper.classes()).toEqual([
      'cdr-layout',
      'cdr-layout--column-gap-one-x',
      'cdr-layout--row-gap-one-x',
      'cdr-layout--columns',
    ]);
    expect(wrapper.element).toMatchSnapshot();
  });
});
