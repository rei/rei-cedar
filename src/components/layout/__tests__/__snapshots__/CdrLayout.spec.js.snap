// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CdrLayout > base classes only 1`] = `
<span
  class="cdr-layout"
>
  <!-- @slot Where all default content should be placed. -->
  
  
</span>
`;

exports[`CdrLayout > basic columns 1`] = `
<div
  class="cdr-layout cdr-layout--columns"
  style="--cdr-layout-columns: 1fr 1fr;"
>
  <!-- @slot Where all default content should be placed. -->
  
  
</div>
`;

exports[`CdrLayout > container query columns 1`] = `
<div
  class="cdr-layout cdr-layout--columns cdr-layout--columns-cq"
  style="--cdr-layout-columns-xs: 1fr; --cdr-layout-columns-sm: 100px 1fr; --cdr-layout-columns-md: 400px 1fr; --cdr-layout-columns-lg: 800px 1fr;"
>
  <!-- @slot Where all default content should be placed. -->
  
  
</div>
`;

exports[`CdrLayout > default snapshot 1`] = `
<div
  class="cdr-layout"
>
  <!-- @slot Where all default content should be placed. -->
  
  
</div>
`;

exports[`CdrLayout > different gap 1`] = `
<div
  class="cdr-layout cdr-layout--column-gap-one-x cdr-layout--row-gap-one-x cdr-layout--columns"
  style="--cdr-layout-columns: 1fr 1fr;"
>
  <!-- @slot Where all default content should be placed. -->
  
  
</div>
`;

exports[`CdrLayout > flow to generate columns 1`] = `
<div
  class="cdr-layout cdr-layout--flow-column"
  style="--cdr-layout-flow-value: 100px;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
  
</div>
`;

exports[`CdrLayout > media query columns 1`] = `
<div
  class="cdr-layout cdr-layout--columns cdr-layout--columns-mq"
  style="--cdr-layout-columns-xs: 1fr; --cdr-layout-columns-sm: 100px 1fr; --cdr-layout-columns-md: 400px 1fr; --cdr-layout-columns-lg: 800px 1fr;"
>
  <!-- @slot Where all default content should be placed. -->
  
  
</div>
`;

exports[`CdrLayout > ratio columns 1`] = `
<div
  class="cdr-layout cdr-layout--columns"
  style="--cdr-layout-columns: 2fr 1fr;"
>
  <!-- @slot Where all default content should be placed. -->
  
  
</div>
`;

exports[`CdrLayout > same gap 1`] = `
<div
  class="cdr-layout cdr-layout--gap-one-x cdr-layout--columns"
  style="--cdr-layout-columns: 1fr 1fr;"
>
  <!-- @slot Where all default content should be placed. -->
  
  
</div>
`;

exports[`CdrLayout > shows default slot 1`] = `
<div
  class="cdr-layout"
>
  <!-- @slot Where all default content should be placed. -->
  
  Default slot
  
</div>
`;
